developer/services/qrexec.md

```diff
@@ -0,0 +1,307 @@
+For instance, when a user selects an application from the KDE menu, it should start in the selected VM.
+Also, it is often useful to be able to pass stdin/stdout/stderr from an application running in a VM to dom0 (and the other way around).
+(For example, so that a VM can notify dom0 that there are updates available for it).
+By default, Qubes allows VMs initiate such communications in specific circumstances.
+The qrexec framework generalizes this process by providing a remote procedure call (RPC) protocol for the Qubes architecture.
+It allows users and developers to use and design secure inter-VM tools.
+
+## Qrexec basics: architecture and examples
+
+Qrexec is built on top of *vchan*, a Xen library providing data links between VMs.
+During domain creation, a process named `qrexec-daemon` is started in dom0, and a process named `qrexec-agent` is started in the VM.
+They are connected over a **vchan** channel.
+`qrexec-daemon` listens for connections from a dom0 utility named `qrexec-client`.
+Let's say we want to start a process (call it `VMprocess`) in a VM (`someVM`).
+Typically, the first thing that a `qrexec-client` instance does is to send a request to the `qrexec-daemon`, which in turn relays it to `qrexec-agent` running in `someVM`.
+`qrexec-daemon` assigns unique vchan connection details and sends them to both `qrexec-client` (in dom0) and `qrexec-agent` (in `someVM`).
+`qrexec-client` starts a vchan server, which `qrexec-agent` then connects to.
+Once this channel is established, stdin/stdout/stderr from the VMprocess is passed between `qrexec-agent` and the `qrexec-client` process.
+
+![qrexec basics diagram](/attachment/wiki/qrexec3/qrexec3-basics.png)
+
+The `qrexec-client` command is used to make connections to VMs from dom0.
+For example, the following command
+
+    $ qrexec-client -e -d someVM user:'touch hello-world.txt'
+
+creates an empty file called `hello-world.txt` in the home folder of `someVM`.
+
+The string before the colon specifies what user to run the command as.
+The `-e` flag tells `qrexec-client` to exit immediately after sending the execution request and receiving a status code from `qrexec-agent` (whether the process creation succeeded).
+With this option, no further data is passed between the domains.
+By contrast, the following command demonstrates an open channel between dom0 and someVM (in this case, a remote shell):
+
+    $ qrexec-client -d someVM user:bash
+
+The `qvm-run` command is heavily based on `qrexec-client`.
+It also takes care of additional activities, e.g. starting the domain if it is not up yet and starting the GUI daemon.
+Thus, it is usually more convenient to use `qvm-run`.
+
+There can be almost arbitrary number of `qrexec-client` processes for a domain (so, connected to the same `qrexec-daemon`, same domain) -- their data is multiplexed independently.
+Number of available vchan channels is the limiting factor here, it depends on the underlying hypervisor.
```

Comment:
hypervisor and OS within domain. For example Windows has much lower limit (about 20).